net/dns/manager.go

```diff
@@ -79,6 +84,23 @@ func NewManager(logf logger.Logf, oscfg OSConfigurator, health *health.Tracker,
 		knobs:    knobs,
 		goos:     goos,
 	}
+
+	// Rate limit our attempts to correct our DNS configuration.  The recovery func is triggered
+	// by SERVFAIL errors, which can be occurring in batches at relatively high frequency.
+	limiter := rate.NewLimiter(1.0/5.0, 1)
+
+	// This will recompile the DNS config, which in turn will requery the system
+	// DNS settings.
+	m.resolver.SetServFailRecovery(func() {
+		if m.config == nil {
+			return
+		}
+		if limiter.Allow() {
+			m.logf("DNS resolution failed due to missing upstream nameservers.  Recompiling DNS configuration.")
```

Comment:
cool idea for a v2: add a client_metric to see how prevalent this is over time

Reply:
In corp, if we make a [] nameservers result an error, then we get a health warning notification BTW...  The whole crux of this is that we have no way of telling if [] is really an error, or if we were just too quick with our network setup.

If we waited until NetworkMonitor called us back with a 'viable' interface, we could probably get around this (just based on the ordering I see in the logs) - but that requires a considerable amount of swift->go plumbing that is very apple specific.